packages/panna-sdk/src/core/utils/activity.ts

```diff
@@ -0,0 +1,454 @@
+          const erc20Tx = tx.token_transfers.find(
+            (e) => e.token.type.toLowerCase() === TokenERC.ERC20
+          );
+
+          amount = {
+            type: TokenERC.ERC20,
+            value: tx.value,
+            tokenInfo: {
+              address: erc20Tx?.token.address,
+              name: erc20Tx?.token.name,
+              symbol: erc20Tx?.token.symbol,
+              decimals: erc20Tx?.token.decimals,
+              type: TokenERC.ERC20,
+              icon: erc20Tx?.token.icon_url
+            }
+          } as ERC20Amount;
+          break;
+        }
+
+        case TokenERC.ERC721: {
+          const erc721Tx = tx.token_transfers.find(
+            (e) => e.token.type.toLowerCase() === TokenERC.ERC721
+          );
+          const erc721TxTotal = erc721Tx?.total as BlockscoutTotalERC721;
+
+          amount = {
+            type: TokenERC.ERC721,
+            tokenId: erc721TxTotal.token_id,
+            instance: {
+              id: erc721TxTotal.token_instance.id,
+              isUnique: erc721TxTotal.token_instance.is_unique,
+              owner: erc721TxTotal.token_instance.owner,
+              tokenInfo: {
+                address: erc721Tx?.token.address,
+                name: erc721Tx?.token.name,
+                symbol: erc721Tx?.token.symbol,
+                decimals: erc721Tx?.token.decimals,
+                type: TokenERC.ERC721,
+                icon: erc721Tx?.token.icon_url
+              }
+            }
+          } as unknown as ERC721Amount;
+          break;
+        }
+
+        case TokenERC.ERC1155: {
+          const erc1155Tx = tx.token_transfers.find(
+            (e) => e.token.type.toLowerCase() === TokenERC.ERC1155
+          );
+          const erc1155TxTotal = erc1155Tx?.total as BlockscoutTotalERC1155;
+
+          amount = {
+            type: TokenERC.ERC1155,
+            tokenId: erc1155TxTotal.token_id,
+            value: erc1155TxTotal.value,
+            instance: {
+              id: erc1155TxTotal.token_instance.id,
+              isUnique: erc1155TxTotal.token_instance.is_unique,
+              owner: erc1155TxTotal.token_instance.owner,
+              tokenInfo: {
+                address: erc1155Tx?.token.address,
+                name: erc1155Tx?.token.name,
+                symbol: erc1155Tx?.token.symbol,
+                decimals: erc1155Tx?.token.decimals,
+                type: TokenERC.ERC1155,
+                icon: erc1155Tx?.token.icon_url
+              }
+            }
+          } as unknown as ERC1155Amount;
+          break;
+        }
+      }
+
+      const activity: Activity = {
+        activityType,
+        transactionID,
+        amount,
+        status
+      };
+      return activity;
+    });
+
+  const metadata: ActivityMetadata = {
+    count: activities.length,
+    offset,
+    limit
+  };
+
+  const result: GetActivityResult = { activities, metadata };
+
+  return result;
+};
+
+export const fillTokenTransactions = async (
```

Comment:
maybe better to have docs here to explain a bit about the function since its not that straightforward

Reply:
I think the func is complex and it would be nice if we can simplify it into points below and add comments,
- Fetch all cached data at the start
- Fetch all token transfers in a single loop, if the cache is incomplete
  - Update cache for new retrieved values
- Process transactions by filling token transfers

we can avoid recursion as it could get hard to debug if there are large number of transactions failing, risk of stack overflow, etc we can use while/for loop instead. Also simplifiers cache usage scattered across the function

Reply:
Refactored the code to be more readable.